### ISSUE TITLE
Data files which are still useful are mistakenly cleaned up when trying to expire a specified snapshot

### ISSUE THREAD

**Issue by hantangwangd** [apache/iceberg#10982]
### Apache Iceberg version

1.6.0 (latest release)

### Query engine

None

### Please describe the bug 🐞

When using the policy of `IncrementalFileCleanup` to expire a specified snapshot, the data files that deleted by this snapshot will be physically deleted from the disk. However, these data files may be referenced by some earlier snapshots. So after that, when we want to use these earlier snapshots to do a `Time Travel` query, we will encounter the problem that some data files do not exist.

For example, if we have a table with actions as follows:

```
table.newAppend().appendFile(FILE_A).commit;	// snapshotID1
table.newDelete().deleteFile(FILE_A).commit;	// snapshotID2
table.newAppend().appendFile(FILE_A2).commit;	// snapshotID3
```

Now we want to expire the specified snapshot `snapshotID2` using the policy of `IncrementalFileCleanup`:

```
((RemoveSnapshots) table.expireSnapshots())
        .withIncrementalCleanup(true)
        .expireSnapshotId(snapshotID2)
        .cleanExpiredFiles(true)
        .commit();
```

After that, we will find `FILE_A` has been physically deleted from the disk. When we want to do a `Time Travel` query on `snapshotID1`, we will fail because `FILE_A` do not exist any more.

### Willingness to contribute

- [ ] I can contribute a fix for this bug independently
- [X] I would be willing to contribute a fix for this bug with guidance from the Iceberg community
- [ ] I cannot contribute a fix for this bug at this time

**Comment by hantangwangd** [comment]
As I understand, we should not directly clean a file in a DELETE entry of the expired snapshot, but can only clean it safely when one of the following conditions are met:

 - There are no Snapshots older than this expired one
 - There are no Snapshots exist between the one that added it and this one

We can use the DELETE entry's file sequence number to figure out the snapshot which added this file to the table. If unable to obtain this message (as in format v1), we should leave the file as it is rather than clean it eagerly.

The fix PR is here: #10983 . If there are any misundertandings, please let me know. Any informations would be greatly appreciated!

**Comment by hantangwangd** [comment]
For reference, thanks for the guidance and feedback from reviewers @rdblue @amogh-jahagirdar @RussellSpitzer, we have ultimately adopted the following way to address this issue:

 - If expireSnapshotId is called and incrementalCleanup is set explicitly to true, we fail before triggering the file cleanup.
 - If expireSnapshotId is called and incrementalCleanup isn't specified, we set incrementalCleanup to false so that reachable cleanup will be used.

In this solution, we accept that incremental cleanup has limitations and from an API perspective ensure that users will not easily fall into a state where they cleanup files that are still referenced.
